src/librustc_builtin_macros/asm.rs

```diff
@@ -0,0 +1,527 @@
+use fmt_macros as parse;
+
+use rustc_ast::ast;
+use rustc_ast::ptr::P;
+use rustc_ast::token;
+use rustc_ast::tokenstream::TokenStream;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_errors::{Applicability, DiagnosticBuilder};
+use rustc_expand::base::{self, *};
+use rustc_parse::parser::Parser;
+use rustc_span::symbol::{kw, sym, Symbol};
+use rustc_span::{InnerSpan, Span};
+use rustc_target::asm::{InlineAsmOptions, InlineAsmTemplatePiece};
+
+struct AsmArgs {
+    template: P<ast::Expr>,
+    operands: Vec<(ast::InlineAsmOperand, Span)>,
+    named_args: FxHashMap<Symbol, usize>,
+    reg_args: FxHashSet<usize>,
+    options: InlineAsmOptions,
+    options_span: Option<Span>,
+}
+
+fn parse_args<'a>(
+    ecx: &mut ExtCtxt<'a>,
+    sp: Span,
+    tts: TokenStream,
+) -> Result<AsmArgs, DiagnosticBuilder<'a>> {
+    let mut p = ecx.new_parser_from_tts(tts);
+
+    if p.token == token::Eof {
+        return Err(ecx.struct_span_err(sp, "requires at least a template string argument"));
+    }
+
+    // Detect use of the legacy llvm_asm! syntax (which used to be called asm!)
+    if p.look_ahead(1, |t| *t == token::Colon || *t == token::ModSep) {
+        let mut err = ecx.struct_span_err(sp, "legacy asm! syntax is no longer supported");
+
+        // Find the span of the "asm!" so that we can offer an automatic suggestion
+        let asm_span = sp.from_inner(InnerSpan::new(0, 4));
+        if let Ok(s) = ecx.source_map().span_to_snippet(asm_span) {
+            if s == "asm!" {
+                err.span_suggestion(
+                    asm_span,
+                    "replace with",
+                    "llvm_asm!".into(),
+                    Applicability::MachineApplicable,
+                );
+            }
+        }
+        return Err(err);
+    }
+
+    let template = p.parse_expr()?;
+    let mut args = AsmArgs {
+        template,
+        operands: vec![],
+        named_args: FxHashMap::default(),
+        reg_args: FxHashSet::default(),
+        options: InlineAsmOptions::empty(),
+        options_span: None,
+    };
+
+    let mut first = true;
+    while p.token != token::Eof {
+        if !p.eat(&token::Comma) {
+            if first {
+                // After `asm!(""` we always expect *only* a comma...
+                let mut err = ecx.struct_span_err(p.token.span, "expected token: `,`");
+                err.span_label(p.token.span, "expected `,`");
+                p.maybe_annotate_with_ascription(&mut err, false);
+                return Err(err);
+            } else {
+                // ...after that delegate to `expect` to also include the other expected tokens.
+                return Err(p.expect(&token::Comma).err().unwrap());
+            }
+        }
+        first = false;
+        if p.token == token::Eof {
```

Comment:
The RFC specifies the `asm!` grammar and in it trailing commas are not permitted.

```
 asm := "asm!(" format_string *("," [ident "="] operand) ["," options] ")" 
```

Either the implementation or the RFC have to be adjusted.